src/dmd/link.d

```diff
@@ -562,59 +562,148 @@ public int runLINK()
+        }
+
+        /**
+        Checks if a C string is null or an emtpy string
+        Params:
+            p = the C string to check
+        Returns:
+            `true` if `p` is null or an empty string
+        */
+        static bool isNullOrEmpty(const(char)* p)
+        {
+            return !p || !p[0];
+        }
+
+        /**
+        Checks if a C string is equal to the string "-Xlinker"
+        Params:
+            p = the C string to check
+        Returns:
+            `true` if `p` is equal to "-Xlinker"
+        */
+        static bool isXlinker(const(char)* p)
+        {
+            return !isNullOrEmpty(p) && (strcmp("-Xlinker", p) == 0);
+        }
+
         // return true if flagp should be ordered in with the library flags
-        static bool flagIsLibraryRelated(const char* flagp)
+        static bool flagIsLibraryRelated(const char* p)
         {
-            const flag = flagp[0 .. strlen(flagp)];
-            bool startsWith(string needle)
-            {
-                return flag.length >= needle.length && flag[0 .. needle.length] == needle;
-            }
+            const flag = p[0 .. strlen(p)];
 
-            return startsWith("-l") || startsWith("-L")
+            return startsWith(p, "-l") || startsWith(p, "-L")
                 || flag == "-(" || flag == "-)"
                 || flag == "--start-group" || flag == "--end-group"
-                || FileName.equalsExt(flagp, "a")
+                || FileName.equalsExt(p, "a")
             ;
         }
+
         /* Add libraries. The order of libraries passed is:
-         *  1. link switches others than with -L command line switch,
+         *  1. link switches without a -L prefix,
                e.g. --whole-archive "lib.a" --no-whole-archive     (global.params.linkswitches)
          *  2. static libraries ending with *.a     (global.params.libfiles)
-         *  3. link switches passed with -L command line switch  (global.params.linkswitches)
+         *  3. link switches with a -L prefix  (global.params.linkswitches)
          *  4. libraries specified by pragma(lib), which were appended
          *     to global.params.libfiles. These are prefixed with "-l"
          *  5. dynamic libraries passed to the command line (global.params.dllfiles)
          *  6. standard libraries.
          */
-        foreach (p; global.params.linkswitches)
+
+        // STEP 1
+        if (global.params.noXlinker) // See https://issues.dlang.org/show_bug.cgi?id=6952
         {
-            if (p && p[0] && !flagIsLibraryRelated(p))
+            const(char)* prior;
+            foreach (p; global.params.linkswitches)
             {
-                argv.push("-Xlinker");
-                argv.push(p);
+                // link switches with a prepended "-Xlinker" should be
+                // be pushed together to `linkerArgs`
+
+                if (!isXlinker(p))
+                {
+                    if (!isNullOrEmpty(p) && !flagIsLibraryRelated(p))
```

Comment:
4 instances of this logic appear. I suggest folding isNullOrEmpty into flagIsLibraryRelated.

Reply:
Addressed in https://github.com/dlang/dmd/pull/10341